### DISCUSSION TITLE
Pyright in a web worker requires refactor

### DISCUSSION THREAD

**Discussion by nchen63** [microsoft/pyright#3222] [Ideas]
I have Pyright working in a web worker, but it required a few changes because there are dependencies on node. Would you be open to refactoring the code to abstract out the node dependencies? I could take a shot at a PR if you're ok with that.

**Comment by erictraut** [comment]
Can you summarize the changes here? The pylance team already has pyright running in a web browser, so I'm surprised to hear that additional refactoring was required to make this work.

**Reply by nchen63** [reply]
The issues I'm running into:

1.  references "process"
2. references "worker_threads"
3. references to "fs" such as Dirent, ReadStream, WriteStream, existsSync, realpathSync

If pylance uses pyright as-is, then maybe I'm taking the wrong approach.

**Reply by nchen63** [reply]
I suspect that these calls are shimmed by Webpack. I'm not sure if that's an ideal solution, though

**Reply by erictraut** [reply]
`worker_threads` is used only if background analysis is enabled. That can be disabled by creating a subclass of `LanguageServerBase` that overrides `createBackgroundAnalysis` to return undefined.

All of the file system calls are handled through an abstract file system interface that can be swapped out for another implementation. So that should require no refactoring — just an implementation of the file system interface that works within a web browser.

That just leaves the global object `process`. That can be easily shimmed by adding a global object that provides dummy implementations of the few methods and varaibles used by pyright: `memoryUsage` and `execArgv`.

So I think it should be straightforward to run pyright in a web browser without any further refactoring or abstractions.

**Reply by nchen63** [reply]
Clearly, it's possible to do these things as workarounds, but needing to shim and use global variables isn't a great experience for a module, and a hard sell for larger projects. 

The change can be as simple as adding a single level of indirection for these dependencies.

**Reply by erictraut** [reply]
All of these are single levels of indirection. We're not interested in adding yet another level of indirection.

**Reply by erictraut** [reply]
The file system and background analysis abstractions are already overridable through well-designed mechanisms. The shim for `process` is admittedly not as clean. We could add a formal abstraction here, but I don't think it's worth it. A shim for `process` involves about five lines of code.

**Reply by wch** [reply]
I've made a version of pyright where it works completely in the browser, at https://github.com/posit-dev/pyright. It was originally based on this [fork by micro:bit](https://github.com/microbit-foundation/pyright/) but is now substantially different.

I mostly followed @erictraut's advice, except that I was able to enable background analysis by making a shim for `node:worker_threads`, which is implemented using Web Workers.

For more information, see https://github.com/posit-dev/pyright/blob/pyright-browser/THIS_FORK.md

**Comment by FrankFang** [comment]
> The pylance team already has pyright running in a web browser

Is there a demo or link?

**Reply by rchiodo** [reply]
You can use it at https://vscode.dev if you install the Pylance extension.
